kevm-pyk/src/kevm_pyk/kevm.py

```diff
@@ -155,13 +155,12 @@ def custom_step(self, cterm: CTerm) -> KCFGExtendResult | None:
         """Given a CTerm, update the JUMPDESTS_CELL and PROGRAM_CELL if the rule 'EVM.program.load' is at the top of the K_CELL.
 
         :param cterm: CTerm of a proof node.
-        :type cterm: CTerm
         :return: If the K_CELL matches the load_pattern, a Step with depth 1 is returned together with the new configuration, also registering that the `EVM.program.load` rule has been applied. Otherwise, None is returned.
-        :rtype: KCFGExtendResult | None
         """
-        load_pattern = KSequence([KApply('loadProgram', KVariable('###BYTECODE')), KVariable('###CONTINUATION')])
-        subst = load_pattern.match(cterm.cell('K_CELL'))
-        if subst is not None:
+        if self.can_make_custom_step(cterm):
+            load_pattern = KSequence([KApply('loadProgram', KVariable('###BYTECODE')), KVariable('###CONTINUATION')])
+            subst = load_pattern.match(cterm.cell('K_CELL'))
```

Comment:
@tothtamas28 @PetarMax Not sure how to integrate the new `can_make_custom_step` without duplicating the logic here.

Reply:
Yes, hmm, that is a bit annoying.

Reply:
One way I could avoid the code duplication and the double execution of the pattern matching would be to have a `_cached_subst` as part of `KEVMSemantics`. Then, `can_make_custom_step` would cache a substitution if it's valid. Finally, in `custom_step`, I would use the `self._cached_subst`.